staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go

```diff
@@ -101,62 +119,200 @@ func WithAudit(handler http.Handler, requestContextMapper request.RequestContext
 		}
 
-		username := "<none>"
-		groups := "<none>"
-		if attribs.GetUser() != nil {
-			username = attribs.GetUser().GetName()
-			if userGroups := attribs.GetUser().GetGroups(); len(userGroups) > 0 {
-				groups = auditStringSlice(userGroups)
-			}
+		ev := &auditapi.Event{
+			Level:     auditapi.LevelResponse,
+			Timestamp: metav1.NewTime(time.Now()),
+			AuditID:   types.UID(id),
 		}
-		asuser := req.Header.Get(authenticationapi.ImpersonateUserHeader)
-		if len(asuser) == 0 {
-			asuser = "<self>"
+		ctx = request.WithAuditEvent(ctx, ev)
+		if err := requestContextMapper.Update(req, ctx); err != nil {
+			glog.Errorf("Failed to attach audit event to the context: %v", err)
+			responsewriters.InternalError(w, req, errors.New("failed to attach audit event"))
+			return
 		}
-		asgroups := "<lookup>"
-		requestedGroups := req.Header[authenticationapi.ImpersonateGroupHeader]
-		if len(requestedGroups) > 0 {
-			asgroups = auditStringSlice(requestedGroups)
+
+		if err := updateEventFromRequest(ev, req, ctx); err != nil {
+			glog.Errorf("Failed to complete audit event from request: %v", err)
+			responsewriters.InternalError(w, req, errors.New("failed to complete audit event from request"))
+			return
 		}
-		namespace := attribs.GetNamespace()
-		if len(namespace) == 0 {
-			namespace = "<none>"
+
+		// we have two code path:
+		//
+		// a) short running request with one event:
+		//
+		//    Wait synchronously for the handler to finish and send a complete event.
+		//
+		// b) long running requests with two events:
+		//
+		//    Spawn a go routine waiting for the status code, synchronously send a first
+		//    request audit event and then give back control to the handler.
+		//    Then continue with a).
+		var firstEventSentCh chan struct{}
+		longRunning := false
+		if longRunningCheck != nil {
+			ri, _ := request.RequestInfoFrom(ctx)
+			if longRunning = longRunningCheck(req, ri); longRunning {
+				firstEventSentCh = make(chan struct{}, 0)
+				defer func() {
+					close(firstEventSentCh)
+				}()
+			}
 		}
-		id := uuid.NewRandom().String()
+		respWriter, codeCh := decorateResponseWriter(w, firstEventSentCh)
 
-		line := fmt.Sprintf("%s AUDIT: id=%q ip=%q method=%q user=%q groups=%q as=%q asgroups=%q namespace=%q uri=%q\n",
-			time.Now().Format(time.RFC3339Nano), id, utilnet.GetClientIP(req), req.Method, username, groups, asuser, asgroups, namespace, req.URL)
-		if _, err := fmt.Fprint(out, line); err != nil {
-			glog.Errorf("Unable to write audit log: %s, the error is: %v", line, err)
+		// for long-running requests send a first event when the status code arrives
+		if longRunning {
+			go func() {
+				defer close(firstEventSentCh)
+
+				code, ok := <-codeCh
+				if !ok {
+					// closed channel
+					return
+				}
+
+				// here the handler is waiting for us synchronously
+
+				if err := backend.LogAuditEvent(ev); err != nil {
+					glog.Errorf(err.Error())
+				}
+
+				// set the code in the second event.
```

Comment:
I would prefer something outside of the status field for that. We use a payload field for meta information otherwise.